ql/instruments/nonstandardswaption.hpp

```diff
@@ -32,62 +32,109 @@
 #include <ql/termstructures/volatility/swaption/swaptionvolstructure.hpp>
 #include <ql/models/calibrationhelper.hpp>
 #include <ql/utilities/disposable.hpp>
+#include <ql/experimental/fx/proxyinstrument.hpp>
 
 namespace QuantLib {
 
-    //! nonstandard swaption class
-    /*! \ingroup instruments
-    */
-
-    class NonstandardSwaption : public Option {
-      public:
-        class arguments;
-        class engine;
-        NonstandardSwaption(const Swaption &fromSwaption);
-        NonstandardSwaption(const boost::shared_ptr<NonstandardSwap> &swap,
-                            const boost::shared_ptr<Exercise> &exercise,
-                            Settlement::Type delivery = Settlement::Physical);
-
-        //! \name Instrument interface
-        //@{
-        bool isExpired() const;
-        void setupArguments(PricingEngine::arguments *) const;
-        //@}
-        //! \name Inspectors
-        //@{
-        VanillaSwap::Type type() const { return swap_->type(); }
-
-        const boost::shared_ptr<NonstandardSwap> &underlyingSwap() const {
-            return swap_;
-        }
-        //@}
-        Disposable<std::vector<boost::shared_ptr<CalibrationHelper> > >
-        calibrationBasket(
-            boost::shared_ptr<SwapIndex> standardSwapBase,
-            boost::shared_ptr<SwaptionVolatilityStructure> swaptionVolatility,
-            const BasketGeneratingEngine::CalibrationBasketType basketType =
-                BasketGeneratingEngine::MaturityStrikeByDeltaGamma) const;
-
-      private:
-        // arguments
-        boost::shared_ptr<NonstandardSwap> swap_;
-        Settlement::Type settlementType_;
-    };
+//! nonstandard swaption class
+/*! \ingroup instruments
+*/
 
-    //! %Arguments for nonstandard swaption calculation
-    class NonstandardSwaption::arguments : public NonstandardSwap::arguments,
-                                           public Option::arguments {
-      public:
-        arguments() {}
-        boost::shared_ptr<NonstandardSwap> swap;
-        Settlement::Type settlementType;
-        void validate() const;
+class NonstandardSwaption : public Option, public ProxyInstrument {
+  public:
+    class arguments;
+    class results;
+    class engine;
+
+    //! proxy description
+    struct Proxy : ProxyDescription {
```

Comment:
Yes, you are right. So we would move NonstandardSwaption::Proxy to the
engine that generates this kind of proxy? A proxy on the other hand could
accept several proxies coming from different generating engine and use them
to calculate approximate prices. Is that what you had in mind?

On 25 August 2016 at 13:08, Luigi Ballabio <notifications@github.com> wrote:

> In ql/instruments/nonstandardswaption.hpp
> <https://github.com/lballabio/QuantLib/pull/18#discussion_r76223619>:
>
> > -    //! %Arguments for nonstandard swaption calculation
> > -    class NonstandardSwaption::arguments : public NonstandardSwap::arguments,
> > -                                           public Option::arguments {
> > -      public:
> > -        arguments() {}
> > -        boost::shared_ptr<NonstandardSwap> swap;
> > -        Settlement::Type settlementType;
> > -        void validate() const;
> > +class NonstandardSwaption : public Option, public ProxyInstrument {
> > +  public:
> > +    class arguments;
> > +    class results;
> > +    class engine;
> > +
> > +    //! proxy description
> > +    struct Proxy : ProxyDescription {
>
> Is it possible that different engines might require different proxy
> instruments? (In which case we might move the struct definition to the
> engine?)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lballabio/QuantLib/pull/18/files/bd5b9652c56d43fe028b508463ee44a8ac9e4c0c#r76223619>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACfDrSukx9MZNMoNnvio74eIxYgfIg8hks5qjXfGgaJpZM4G5QbI>
> .
>